### PR TITLE
Remove system-linker flag in iree_bytecode_module

### PR DESCRIPTION
**Pull request by bjacob**

### CODE REVIEW

File: build_tools/cmake/iree_bytecode_module.cmake [hunk 1]
```diff
@@ -108,10 +108,11 @@ function(iree_bytecode_module)
...
   list(APPEND _ARGS "${_MODULE_FILE_NAME}")
 
-  # If an LLVM CPU backend is enabled, supply the linker tool.
+  # If an LLVM CPU backend is enabled, supply the embedded linker tool.
+  # Note: "lld" requires a -flavor flag. This is passed in
+  # EmbeddedLinkerTool.cpp.
   if(IREE_LLD_TARGET)
     iree_get_executable_path(_LINKER_TOOL_EXECUTABLE "lld")
-    list(APPEND _ARGS "-iree-llvm-system-linker-path=\"${_LINKER_TOOL_EXECUTABLE}\"")
     list(APPEND _ARGS "-iree-llvm-embedded-linker-path=\"${_LINKER_TOOL_EXECUTABLE}\"")
```
Comment (ScottTodd) [inline build_tools/cmake/iree_bytecode_module.cmake:116]: Thinking about this more deeply now that I'm more awake - there's more going on here and I don't think a simple patch to the flags here is the right general fix.

TLDR: I think I'll put together a PR that approaches this in a different way

---

We should maintain consistency between `iree_bytecode_module.bzl` and this file, or call out why they are different.

https://github.com/google/iree/blob/9a4e23e3fe06aa53e89d650c2fe974615d52d243/build_tools/bazel/iree_bytecode_module.bzl#L77-L78

So I see a few paths forward here that keep them consistent:

1. Add a `LINKER_TOOL` argument here to the CMake function and set up the bazel_to_cmake conversion. Change the tool from the default where needed (e.g. where `-iree-llvm-link-embedded=false` is set)

2. Remove the `linker_tool` argument from the Bazel function and add comments to the files explaining:
    * The CMake build directory is searched for tools automatically, so no explicit flag is needed
    * The Bazel build directory is _not_ searched, so explicit flags are needed

---

Stepping back a bit, the issue you hit was using `-iree-llvm-link-embedded=false` on Linux, which uses [UnixLinkerTool.cpp](https://github.com/google/iree/blob/main/iree/compiler/Dialect/HAL/Target/LLVM/internal/UnixLinkerTool.cpp). UnixLinkerTool, unlike EmbeddedLinkerTool and WasmLinkerTool, uses ld instead of lld and does not pass lld's `-flavor` flag. Assuming we want to keep that UnixLinkerTool behavior, we should not be passing lld explicitly as the system linker. That's especially true on Windows, though we don't support Windows Bazel now :P.

The similar issue I hit was using the WasmLinkerTool, which should be finding `wasm-ld` or `lld`.

So maybe:

* `iree_bytecode_module` (Bazel and CMake) uses that go through UnixLinkerTool/WindowsLinkerTool should leave `-iree-llvm-system-linker-path` unset and rely purely on the C++ tool discovery
* uses that go through WasmLinkerTool should set `-iree-llvm-system-linker-path` to lld (or we could add a new `-iree-llvm-wasm-linker-path` and always set it? that actually seems best)
* `-iree-llvm-embedded-linker-path` should still be set

**Review by ScottTodd — CHANGES_REQUESTED** [review]

File: build_tools/cmake/iree_bytecode_module.cmake (hunk 1 above)
Comment (bjacob) [reply build_tools/cmake/iree_bytecode_module.cmake:116]: OK I'll cancel review here and wait for your own PR.  I'm happy with anything that resolves my immediate issue here, which (as you noted above) is I am (due to TSan) using `-iree-llvm-link-embedded=false` on Linux, so I'm using the system linker, and at the moment `iree_bytecode_module` specifies it explicitly as `lld`, so we are executing `lld`, but as we're using `UnixLinkerTool`, which does not pass a `-flavor` flag, that fails. In other words, explicitly passing `lld` as the system linker defeats an assumption that `UnixLinkerTool` makes.
